node/src/syncer.rs

```diff
@@ -618,12 +613,28 @@ where
 
     let network_head = p2p.get_head_header().await?;
 
-    // Insert HEAD to the store and initialize header-sub
+    // Insert HEAD to the store and initialize header-sub.
+    // This will apply insertion restrictions.
     store.insert(network_head.clone()).await?;
 
     Ok(network_head)
 }
 
+async fn header_sub_recv(
+    rx: &mut Option<mpsc::Receiver<ExtendedHeader>>,
+) -> Result<ExtendedHeader> {
+    poll_fn(|cx| {
+        let rx = rx.as_mut().expect("header-sub not initialized");
+
+        match rx.poll_recv(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Some(header)) => Poll::Ready(Ok(header)),
+            Poll::Ready(None) => Poll::Ready(Err(SyncerError::P2p(P2pError::WorkerDied))),
+        }
+    })
+    .await
```

Comment:
I don't understand, why this cannot be
```rs
rx.recv().await.ok_or(...)
```

Reply:
When I was writing this, its first implementation was returning `Poll::Pending` on closed channel instead of an error. At the time I didn't thought to change it to `ok_or` when I changed it to return an error.

Reply:
switched to `ok_or`